cpp/src/dictionary/replace.cu

```diff
@@ -0,0 +1,242 @@
+ * @brief This tility uses `copy_if_else` to replace null entries using the input bitmask as a
+ * predicate.
+ *
+ * The predicate identifies which column row to copy from and the bitmask specifies which rows
+ * are null. Since the `copy_if_else` accepts iterators, we also supply it with pair-iterators
+ * created from indexalators and the validity masks.
+ *
+ * @tparam ReplacementItr must be a pair iterator of (index,valid).
+ *
+ * @param input lhs for `copy_if_else`
+ * @param replacement_iter rhs for `copy_if_else`
+ * @param mr Device memory resource used to allocate the returned column's device memory.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @return Always returns column of type INT32 (size_type)
+ */
+template <typename ReplacementIter>
+std::unique_ptr<column> replace_indices(column_view const& input,
+                                        ReplacementIter replacement_iter,
+                                        rmm::mr::device_memory_resource* mr,
+                                        cudaStream_t stream)
+{
+  auto const input_view = column_device_view::create(input, stream);
+  auto const d_input    = *input_view;
+  auto predicate        = [d_input] __device__(auto i) { return d_input.is_valid(i); };
+
+  auto input_pair_iterator = make_nullable_index_iterator<true>(input);
+  return cudf::detail::copy_if_else(true,
+                                    input_pair_iterator,
+                                    input_pair_iterator + input.size(),
+                                    replacement_iter,
+                                    predicate,
+                                    mr,
+                                    stream);
+}
+
+}  // namespace
+
+/**
+ * @copydoc cudf::dictionary::detail::replace_nulls(cudf::column_view const&,cudf::column_view
+ * const&,rmm::mr::device_memory_resource*,cudaStream_t)
+ */
+std::unique_ptr<column> replace_nulls(dictionary_column_view const& input,
+                                      dictionary_column_view const& replacement,
+                                      rmm::mr::device_memory_resource* mr,
+                                      cudaStream_t stream)
+{
+  if (input.size() == 0) { return cudf::empty_like(input.parent()); }
+  if (!input.has_nulls()) { return std::make_unique<cudf::column>(input.parent()); }
+  CUDF_EXPECTS(input.keys().type() == replacement.keys().type(), "keys must match");
+  CUDF_EXPECTS(replacement.size() == input.size(), "column sizes must match");
+
+  // first combine the keys so both input dictionaries have the same set
+  auto input_matched    = dictionary::detail::add_keys(input, replacement.keys(), mr, stream);
+  auto const input_view = dictionary_column_view(input_matched->view());
+  auto repl_matched     = dictionary::detail::set_keys(
+    replacement, input_view.keys(), rmm::mr::get_current_device_resource(), stream);
+  auto const repl_view = dictionary_column_view(repl_matched->view());
```

Comment:
You mentioned in a meeting that this is becoming a common pattern. Should it be a detail function for reuse?